source/common/protobuf/utility.cc

```diff
@@ -506,6 +511,150 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+std::unique_ptr<Protobuf::Message> typeUrlToNewMessage(Protobuf::MessageFactory& message_factory,
+                                                       absl::string_view type_url) {
+  const absl::string_view type_name = TypeUtil::typeUrlToDescriptorFullName(type_url);
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          static_cast<std::string>(type_name));
+  return std::unique_ptr<Protobuf::Message>(
+      descriptor == nullptr ? nullptr : message_factory.GetPrototype(descriptor)->New());
+}
+
+// Forward declaration for mutually-recursive helper functions.
+void redact(Protobuf::Message* message, bool ancestor_is_sensitive);
+
+// To redact an `Any`, we have to unpack it to its original type to redact it.
+bool redactAny(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, any->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact ProtobufWkt::Any with unknown type URL {}",
+                     any->type_url());
+      return false;
+    }
+
+    any->UnpackTo(typed_message.get());
+    redact(typed_message.get(), ancestor_is_sensitive);
+    any->PackFrom(*typed_message);
+    return true;
+  }
+  return false;
+}
+
+// To redact a `TypedStruct`, we have to reify it based on its `type_url` to redact it.
+bool redactTypedStruct(Protobuf::Message* message, bool ancestor_is_sensitive) {
+  auto* typed_struct = dynamic_cast<udpa::type::v1::TypedStruct*>(message);
+  if (typed_struct != nullptr) {
+    Protobuf::DynamicMessageFactory message_factory;
+    auto typed_message = typeUrlToNewMessage(message_factory, typed_struct->type_url());
+    if (typed_message == nullptr) {
+      // If the type URL doesn't correspond to a known proto, give up redacting and treat this
+      // message the same as a `ProtobufWkt::Struct`. See the documented limitation on
+      // `MessageUtil::redact()` for more context.
+      ENVOY_LOG_MISC(warn, "Could not redact udpa::type::v1::TypedStruct with unknown type URL {}",
```

Comment:
Does this actually happen? If it happens a lot the warning message could get spammy. If it happens rarely, do we want to ASSERT here to try to track them down? Or does this depend on config?

Reply:
Yeah, there are two situations where this would happen:

* The configuration is simply malformed; the `type_url` is bogus.
* The configuration is ok, but references a message type that isn't registered in the `DescriptorPool`. IIUC the pool knows about all proto message types that were statically compiled into the binary, but doesn't know about dynamically linked extensions.

Reply:
There are no dynamically linked extensions..